src/unix/async.c

```diff
@@ -61,43 +61,18 @@ int uv_async_send(uv_async_t* handle) {
   if (ACCESS_ONCE(int, handle->pending) != 0)
     return 0;
 
-  /* Tell the other thread we're busy with the handle. */
-  if (cmpxchgi(&handle->pending, 0, 1) != 0)
-    return 0;
-
-  /* Wake up the other thread's event loop. */
-  uv__async_send(handle->loop);
+  /* The cmpxchgi may permit another thread to immediately destroy the handle,
+   * so be careful not to dereference it again after that point. */
+  uv_loop_t* loop = ACCESS_ONCE(uv_loop_t*, handle->loop);
 
-  /* Tell the other thread we're done. */
-  if (cmpxchgi(&handle->pending, 1, 2) != 1)
-    abort();
+  if (cmpxchgi(&handle->pending, 0, 1) == 0)
+    uv__async_send(loop);
```

Comment:
This isn't safe. There's a time window between cmpxchgi() and uv__async_send() where the reference to the event loop can go stale. Put another way, you're referencing it outside the critical section. That's exactly the issue that 37042a5bca8b392f71d7da6e01457ece5a148e99 addresses. :-)

The commit log talks about racing with closing the async handle but the same is true for the event loop, of course.

Reply:
Yeah, true. I don't usually care about loop deletion, but it means we need to keep both facets: make sure we don't dereference `handle` after the `cmpxchgi` _and_ move the spin lock to uv_loop_close.